### ISSUE TITLE
Add a traditional block-based editor

### ISSUE THREAD

**Issue by scotws** [scotws/TaliForth2#113]
Traditional Forths used a block-based editor -- see http://kestrelcomputer.github.io/kestrel/2016/03/29/vibe-2.2 for an example, and https://www.youtube.com/watch?v=1bSxUMa_jdw for a demonstration -- that works on blocks (see https://forth-standard.org/standard/block). Since Tali is bound to be used on rather simple systems, and lots of people seem to like these editors, this should at least be offered as an option in addition to the already planed **ed**-like line-based editor (see #16). As with the disassembler, the user should be able to change which editor to use with fairly little trouble. 

To enable this, the BLOCK word set **BLK, BLOCK, BUFFER, FLUSH, LOAD, SAVE-BUFFERS,** and **UPDATE** are required; it would be nice to have the BLOCK EXT wordset as well with **EMPTY-BUFFERS, LIST, SCR,** and **THRU**.

**Comment by SamCoVT** [comment]
Does it make sense to expect the kernel provide an API for getting and storing blocks (similar to R/W for fig forth)?  That way a user of Tali2 would write a kernel driver for their I2C/SPI/SDCARD/IDE/whatever device to supply reading and writing, and Tali2 would do everything else.  I'm interested in getting an I2C flash memory working on my SBC, and I think working in blocks will work best for a simple system.
I'll read up on the block words in the standard and then take a look at the examples you mentioned.  I tried using "simulated" block storage when playing with FIG Forth (using essentially a RAM drive), but I was never able to get it to work well.  That gave me at least a basic understanding of what most of these words do.

**Comment by scotws** [comment]
How very interesting, I would never have thought of that ... sounds like something that would be very useful. -- Just out of curiosity, why I2C and not SPI? From what I've read on 6502.org, there are actually chips that provide SPI services for the 6502 family.

**Comment by SamCoVT** [comment]
The main reason for choosing I2C over SPI (although there's nothing saying you can't have both) is that I already have some I2C ICs and I have enough experience with it that I can troubleshoot it with an oscilloscope without having to reference the documentation.

The SPI bus is certainly faster, but you need a separate enable line for each device.  I also don't like the way the standard doesn't define the clock phase and polarity (giving 4 different setup possibilities,) meaning you may have to re-setup the bus every time you want to talk to a different device.  It's fine for a single device, but can be annoying when you have multiple devices.

When you have more than a couple of devices, I2C ends up being less wiring.  You can also add another device to an I2C bus very easily (as long as it has a unique address) because you don't have to find an extra enable line for it.  I've used an I2C bus with 64 sensors on it distributed along a 5 meter cable, and it worked great for the application.

**Comment by SamCoVT** [comment]
TLDR version: Me rambling about my thoughts for block support.

I have my I2C hardware working now in forth, and I'm able to read/write bytes.  The CTRL-P/CTRL-N history was super useful on my hardware for developing this.

I think I have enough working to write some of the BLOCK words.  I wanted to think my way through things, so let me know if you see anything that isn't true or isn't going to work.

I've been reading the spec, and I think I may be able to get away with a single general purpose 1K block buffer + 1 word for the block number + 1 word for the modified flag + 1K interpreting buffer (whose block number will be held in BLK).

The standard says: `A program may reposition the parse area within a block by manipulating >IN.` which tells me that the entire block needs to be in RAM while it is being parsed, so it looks like we won't be able to reuse the 255-byte input buffer we already have.  Also, using a load command shouldn't mess up the rest of the line being interpreted, so interpreting a block will have to have its own 1K buffer.  This makes it so the words in the block can operate on a block.

The blocks are space filled, so I can use INTERPRET to create LOAD by giving it the interpreting buffer address and a size of 1K, essentially treating it like a 1K string.  The issue is that the commands inside the block can load other blocks, so there will have to be a way to save/restore the interpreted block (which might be just the block number and IN>).  I plan to copy the block over into the interpreting buffer and running it from there.  Words like LOAD will have to see if they are already running from a block, and if they are, they will need to restore that block to the interpreting buffer when they are done with the new block (INTERPRET already saves IN> and the address, so perhaps only the text in the interpreting buffer needs to be restored).

If the comment word \ is not updated for blocks, it will not be usable in blocks, however ( and ) should work fine.  I'm OK with that and updating \ is optional for blocks.  We can deal with it after getting basic blocks words working.

**Comment by SamCoVT** [comment]
In reading the specification again, I think I can make this work with only a single block buffer (and associated variables).  In order to make EVALUATE and LOAD nestable, I will need to make sure that the BLK number gets saved and restored (on the return stack along with the other input variables).  When `BLK`, `INPUT-SOURCE` and `>IN` are restored, if BLK is non-zero, BLOCK should be called on it to refill the input buffer with the correct block to continue processing.  SOURCE should be updated with the results of BLOCK to make it compatible with multiple buffers, if we ever do that.

```
\ After an EVALUATE or LOAD have restored BLK, INPUT-SOURCE, and >IN
\ make sure the block we were interpreting from is in memory when we are
\ interpreting out of a block.
blk @ =0 if blk @ block source ! then
```

**Comment by SamCoVT** [comment]
Making progress.  I have the following words working with my 128KB I2C EEPROM:
BLOCK
BUFFER
EMPTY-BUFFERS
FLUSH
SAVE-BUFFERS
UPDATE

I believe I'll actually need an editor now before I can try LOAD and THRU.  I've just been poking bytes into a buffer (or using loops to fill the buffer) and saving them.  With this much written, I'm still fairly confident I can get away with a single 1K buffer.

**Comment by SamCoVT** [comment]
I have LOAD and THRU working now, and I need some help with some decisions.

## Question 1:
I'd like to make the block support easy for others to implement, so I think I'd like to make a word like R/W in FIG forth.  It's defined like this in the FIG glossary:
```
R/W addr blk f --
        The fig-FORTH standard disc read-write linkage. addr specifies the
        source or destination block buffer, blk is the sequential number of
        the referenced block; and f is a flag for f=O write and f=l read.
        R/W determines the location on mass storage, performs the read-write
        and performs any error checking.
```
Does it make sense to add this word (or one like it) with **DEFER**, and possibly even add an example implementation of a RAM drive that provides a few "blocks" to play with in RAM?  That way, the user can use **IS** to redirect R/W to their hardware handling routine.  The RAM drive version can also be used for testing all of the block words.

## Question 2:
How many buffers should we provide?  They are 1K each.  Options include:

A) One, which is what I have now, and it works.
B) Two, which is the minimum recommended by the 6502.org folks when I asked.
C) User decides when bringing in the BLOCK words, which makes the code slightly more complicated, but I don't think overly so.

**Comment by SamCoVT** [comment]
One more thought - if there is more than 1 buffer, then it's possible that a block might start being interpreted in one buffer, but then might be brought back in to another buffer (if the buffers were used while evaluating a string, for example).  The current version of **EVALUATE** assumes it can just put cib back the way it was, and this works for a single buffer (my LOAD word reloads the buffer with the block it had before (if BLK wasn't zero previously.))  

With multiple buffers, cib might have to be changed to the new buffer that has the block.  **>IN** should be fine the way it currently is saved/restored.   I currently don't see any way to modify CIB through forth except by using **!** and a hardcoded address.  Does it make sense to expose cib, or is there a word that does this already?

**Comment by scotws** [comment]
So the way I understand how these blocks work, just to be sure my mental model is correct, is that there are a certain number of buffers in RAM, and a really, really large number of blocks on some mass storage device (HD, SD card, whatever). Those are numbered, and I sort of have to remember which block contains what (or define **CONSTANT**s?). The whole thing functions like virtual memory in the sense that I just say "dude, I want to edit block 13", and the machine handles the details like loading it into memory and saving it later with flushes (which, by the way, means we need to adapt **BYE**). Ideally, I don't need to know anything about how many buffers there are in memory.

(Most of this is based on https://www.complang.tuwien.ac.at/forth/gforth/Docs-html/Blocks.html , BTW.)

Assuming I've understood that (and we might think about a tutorial at some point, I don't think there is a good one out there) --

**R/W** - My understanding is that this is a lower-level word, which is why Gforth doesn't need it -- they actually use the file system as the base, a luxury we don't have. I agree that something like this would make our users' lives a lot easier. Is it worth adding routines like this to the kernel, similar to the way we do text input and output, and then define the vectors in definitions.asm? People who don't use blocks could ignore it / it could be disabled by default, and the nasty low-level details could live in the kernel. We have **INPUT** and **OUTPUT** for redirection, maybe something like **BLOCK-READ** and **BLOCK-WRITE** for blocks? 

(I think what I don't like about **R/W** is that it does too much at once and you have to remember which version of the flag is for reading and which for writing.)

(It would be nice if we could figure out some way for the user to use a fake block system when testing with py65, but that sounds like a whole new level of complexity we might want to save for after we have Version 1 done.)

**Number of buffers** - My suggestion would be to have one buffer at first to get everything up and running, just because it is simpler and easier to test. That would just be a first step, though: I think you are completely right that we want to add the ability to include lots and lots of buffers for people with lots and lots of RAM, so at some point we will have to face the cib problem. I dimly remember that **REFILL** somehow was involved with blocks on some level, but at the time I just put fingers in my ears and went "lalalala" until that part of the specification went away ... 

(Also, I'm wondering about things like shadow blocks (which sound cool, though I'm not sure I understand them completely) and how blocks will work with multitasking.)

**Comment by SamCoVT** [comment]
Your mental model sounds good.  

As a programmer, you just have to remember that if you use a word like **BLOCK** or **BUFFER**, it invalidates any other buffer addresses you were previously using (because you don't know which buffer got re-used).  You'll see that most code calls **BLOCK** again any time it's possible the block may have been swapped out.  There's very little penalty for doing this if the block is actually still in a buffer, as it will just give you the address of that buffer gain. On the other hand, if the block isn't in any of the buffers, Forth will quickly put it into one of them and give you the address of that buffer.  

That's the reason I need access to **CIB** if we go multi-buffer.  I may need to resume interpreting from a different buffer if the block moved.  The offset in **>IN** will be right, but the address in **CIB** will need to be moved.

Here is the sample code given in the FIG Forth installation manual for R/W:
```
RAM DISC SIMULATION

If disc is not available, a simulation of BLOCK and BUFFER may be made
in RAM.  The following definitions setup high memory as mass storage.
Referenced 'screens' are then brought to the 'disc buffer' area.  This
is a good method to test the start-up program even if disc may be
available.

HEX
4000  CONSTANT LO ( START OF BUFFER AREA )
6800  CONSTANT HI ( 10 SCREEN EQUIVALENT )
: R/W >R ( save boolean )
   B/BUF * LO + DUP
   HI > 6 ?ERROR ( range check )
   R> IF ( read ) SWAP ENDIF
   B/BUF CMOVE ;

Insert the code field address of R/W into BLOCK and BUFFER and proceed
as if testing disc.  R/W simulates screens 0 thru 9 when B/BUF is 128,
in the memory area $4000 thru $6BFF.
```
It's always using CMOVE (now named MOVE), but it swaps the addresses around for reading vs writing. I don't think we need 10 screens.  Just screens 1 and 2 are enough to test both LOAD and THRU.

I'm OK with words like **BLOCK_READ** and **BLOCK_WRITE**.  I think it was originally one word because you had to poke the address of your R/W into the existing **BLOCK** and **BUFFER** words to patch FIG Forth.  We now have options like **DEFER**.  If we make them both deferred words in the dictionary, but point them at some memory-copying versions (using memory near the end of RAM) in assembly in the kernel, then the user can either rewrite the assembly versions or they can change the deferred word to a version written in Forth.

In regards to the number of buffers, starting with one sounds like a good plan now that I've proven it's possible.  On the 6502.org forum, one of the users showed off the method he uses in his Forth, which is to allow the user to set the number of buffers and then run a **CONFIGURE** word where it puts the buffers at the end of ram and then moves the "last valid ram address" to just before the buffers.  I think that's a pretty good way of doing it.

The shadow blocks are just blocks used to comment forth code, with the comment on line 7 matching the code in a different block 7.  Some systems use even/odd numbered blocks for this, and some use an offset (eg. block 11 might have the comments for the code in block 1).  Unless our editor supports them (they really only fit well side-by-side on a 132-column screen), it's really just a user practice and we don't need to do anything special to support them.

For multitasking, it's up the user to only have one task at a time using blocks/buffers.

**Comment by SamCoVT** [comment]
#137 is the pull request for the initial block support needed to work on the screen editor.
I also put the code for my current editor into the manual in the Editor chapter.  You should check it out and let me know what you think.  I'm happy to use different word names.  I didn't find a lot of standardization across the few editors I looked at, so I'm currently doing my own thing using **E** to erase a line a **O** to overwrite it.  I also made an **erase-screen** and **enter-screen**.  I use the latter to just copy/paste an entire screen from my editor.

**Comment by SamCoVT** [comment]
If you have a vision of what the screen editor should look like, or what words it should have, now would be a good time to offer your thoughts.  I recently noticed that my E word will interfere with the hex number of the same name, and I probably want to do something about it.  I currently find that I really only need words to erase/change an entire line or an entire screen, so I'm fine with only offering that.  If the user wants more, they can always write it!

I'd like to get the editor into forth_word.fs and then start work on assemblerizing the block things to speed Tali's boot time back up and reduce the ROM usage.  Once the blocks and editor are working, we can create a separate issue for that.

**Comment by scotws** [comment]
Notes on a screen editor while I'm playing with blocks: I think this should be the default built-in editor (called by **EDIT**) because it will be easiest to use (the line-based one is mostly because I like them, but I realized that you have to be seriously weird to use one, even by vim-user standards). The **E** problem you mention will bite lots of people in strange ways sooner so later, I agree, so it would be best if we could avoid that, I think.

There doesn't seem to be any standard for screen editors, which is nice. Probably the easiest system for the user would be if we could use arrow keys to move in the screen and the common `<CTRL>-x`, `<CTRL>-v` bindings everybody knows from Word. This, of course, is something I've been putting off because it honestly sounds like a world of pain to get working in **ACCEPT** whenever I have read about these things on the forum, but I think it would be the most intuitive for the user. Obviously nobody did that back then when blocks were regularly used. Now that Microsoft has trained millions of people to use those keyboard combinations, we'd have a system that should work instantly. 

(We might end up considering a completely separate REPL in assembler for this editor instead of feeding everything through **ACCEPT**. I'm experimenting with a hybrid system for the line editor, and so far am not sure if I like it.)

Once the basics are working, we could see about including other stuff from that keyboard block above the arrow keys (INS/DEL and friends) as well -- Page Down for next block, etc.

**Comment by scotws** [comment]
Looking wonderful, thank you. I made some minor, pretty much cosmetic changes. The one thing that I did differently is rename **BLOCK-RAMDRIVE** to simply **RAMDRIVE**. I'm wondering if we want to have the RAM drive as a generic feature that can also be used separate from the block words for testing hardware routines? You'd be the better judge of that than I would. Then we could rename **BLOCK-RAMDRIVE-INIT** to just **RAMDRIVE-INIT**. 

One thing I noticed while experimenting: **LIST** should probably be built-in, because it was the first thing I tried to do (and the first thing on some of the tutorials). I know it's optional, but I really wanted to _see_ what was in that block :-). 

Once we have the location of **BLK** figured out (#139), we'll have to go back and fix that part of **REFILL** where I was lazy and didn't check it. **REFILL** is one reason why I think we're going to need a tutorial for blocks because there doesn't seem to be much out there about how to use a combination of blocks and ANS Forth. 

All in all, I think I'm becoming a convert to blocks for computers our size. The big advantage I see is that the system is seriously flexible for all the strange stuff people are going to connect while still being easy to use (the difference between **BLOCK** and **BUFFER** is a bit to minimal for my taste, but that's what we're stuck with). Thanks again!

**Comment by scotws** [comment]
@SamCoVT In case it helps you any, it looks like I'm going to have to reserve six bytes of zero page for the line editor, which I'll be naming, uh, `editor1` to `editor3` in two-byte units. I thought I could get by with reusing various stuff, but that's really asking for trouble. Since we can't use two editors at the same time (we'll cross that bridge once we get to multitasking), the block editor should be free to use them as well.

**Comment by SamCoVT** [comment]
I'll be assemblerizing my block editor code and adding it to the EDITOR-WORDLIST.  That will give us a minimal working block editor and we can see where we are on ROM usage.

**Comment by scotws** [comment]
So we can scratch that off the list, thanks - at some point we might want to figure out (if we have the space) how to get an editor that we can use the arrow keys etc with. It seems that requires stuff like "line discipline" in the input, which I think can wait after Version 1.0. Closing this for now.
